### PR TITLE
Using pipeline behavior to manage entity framework dbcontext.

### PR DESCRIPTION
**Pull request by ramonsmits**
A customer was asking how they could inject their dbcontext. I have adjusted the sample so that it uses a custom behavior with the responsibility to call `SaveChanges` and the container configuration so that it contains an `IDbConnection` so it can create the `ReceiveDbContext`.

### CODE REVIEW

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/DbContextBehaviorRegistration.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+#region DbContextBehavior
+class DbContextBehavior<T> : IBehavior<IncomingContext> where T : DbContext
+
+    #endregion
+{
+    public void Invoke(IncomingContext context, Action next)
+    {
+        #region DbContextBehaviorSaveChanges
+        DbContext ctx = context.Builder.Build<T>();
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/DbContextBehaviorRegistration.cs]: Should not use service location

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/DbContextBehaviorRegistration.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        #region DbContextBehaviorPipelineRegistration
+
+        busConfiguration.Pipeline.Register<DbContextBehavior<ReceiverDataContext>.Registration>();
+        
+        #endregion
+    }
+}
+
+#region DbContextBehavior
+class DbContextBehavior<T> : IBehavior<IncomingContext> where T : DbContext
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/DbContextBehaviorRegistration.cs]: In the context of business code a generic adds no value here

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/OrderSubmittedHandler.cs [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
-using System.Data.Common;
-using NServiceBus.Persistence.NHibernate;
 
 public class OrderSubmittedHandler : IHandleMessages<OrderSubmitted>
 {
     public IBus Bus { get; set; }
-    public NHibernateStorageContext StorageContext { get; set; }
+    private readonly ReceiverDataContext ctx;
+
+    public OrderSubmittedHandler(ReceiverDataContext ctx)
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/OrderSubmittedHandler.cs]: Mixing constructor and property injection will break some containers. U need to use one of the other

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/Program.cs [hunk 4]
```diff
@@ -7,6 +7,8 @@ class Program
 {
     static void Main()
     {
+        Console.Title = "Receiver";
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/Program.cs]: While i dont object to this change it now makes it inconsistent with all the other samples. If u want to do this then do do a new pr that just does this and then document it as the recommended  approach for writing samples

File: samples/sqltransport-nhpersistence-outbox-ef/sample.md [hunk 5]
```diff
@@ -13,6 +18,7 @@ related:
 ---
 
  1. Make sure you have SQL Server Express installed and accessible as `.\SQLEXPRESS`. Create two databases: `sender` and `receiver`.
+  * You can also use localdb by replacing `SQLEXPRESS` with `(localdb)\v11.0`.
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/sample.md]: This is assumed knowledge for anyone working with sql. Also the list of ways u can connect to the various sql types is unbound. And if u wanted to do this properly you would need to apply this to every sample using sql. So i would say for now just remove it.

File: samples/sqltransport-nhpersistence-outbox-ef/sample.md [hunk 6]
```diff
@@ -70,4 +76,51 @@ Finally the messages in the outbox are pushed to their destinations. The timeout
 
 ### How it works?
 
-All the data manipulations happen atomically because SQL Server 2008 and later allows multiple (but not overlapping) instances of `SqlConnection` to enlist in one `TransactionScope` without the need to escalate to DTC. The SQL Server manages these transactions like they were one `SqlTransaction`.
+### Database connection sharing
+
+All the data manipulations happen atomically because SQL Server 2008 and later allows multiple (but not overlapping) instances of `SqlConnection` to enlist in one `TransactionScope` without the need to escalate to DTC. The SQL Server manages these transactions like they were one `SqlTransaction`. This does not work for all databases. This implementation shares the same `IDbConnection` between the dependencies that require the connection resulting on all dependencies to read and write to the same database.
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/sample.md]: How do manipulations happen automatically?

File: samples/sqltransport-nhpersistence-outbox-ef/sample.md [hunk 7]
```diff
@@ -70,4 +76,51 @@ Finally the messages in the outbox are pushed to their destinations. The timeout
...
+
+<!-- import DbContextBehaviorPipelineRegistration -->
+
+## Alternatives
+
+An alternative of sharing the `IDbConnection` instance is instructing the container how to create the `ReceiverDataContext` by first resolving the `NHibernateStorageContext` and use that to build it.
+
+<!-- import ReceiverDataContextAlternative -->
+
+The current implementation is usefull when you would like to use the `IDbConnection` with, for example, the [Dapper](https://github.com/StackExchange/dapper-dot-net) micro orm framework.
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/sample.md]: Not really sure how the line about dapper adds value.

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/OrderSubmittedHandler.cs [hunk 8]
```diff
@@ -1,12 +1,15 @@
 using System;
 using NServiceBus;
-using System.Data.Common;
-using NServiceBus.Persistence.NHibernate;
 
 public class OrderSubmittedHandler : IHandleMessages<OrderSubmitted>
 {
     public IBus Bus { get; set; }
-    public NHibernateStorageContext StorageContext { get; set; }
+    private readonly ReceiverDataContext ctx;
```
Comment (SimonCropp) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/OrderSubmittedHandler.cs]: Private and readonly are redundant

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/DbContextBehaviorRegistration.cs (hunk 2 above)
Comment (ramonsmits) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/DbContextBehaviorRegistration.cs]: It does, as you can have multiple db contexts in a single application.

File: samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/OrderSubmittedHandler.cs (hunk 3 above)
Comment (ramonsmits) [inline samples/sqltransport-nhpersistence-outbox-ef/Version_2.1/Receiver/OrderSubmittedHandler.cs]: You are right, currently confusing.

File: samples/sqltransport-nhpersistence-outbox-ef/sample.md (hunk 7 above)
Comment (ramonsmits) [inline samples/sqltransport-nhpersistence-outbox-ef/sample.md]: Does the comment about injecting an `IDbConnection` add value? I added the dapper remark to having more then one persistence framework (ef, nh, dapper, etc) that share the same connection.

**Comment by SimonCropp** [comment]
closing this one. it simply makes that sample too complex. the sample is already mixes the concepts of sqltransport+nhpersistence+outbox+ef. better to just do a sample that shows EF+context injection

**Comment by ramonsmits** [comment]
@SimonCropp A smaller sample for the EF DbContext behavior is possible.

So that would then be for example MSMQ with nh persistence where this EF DbContext behavior is used?

**Pull request closed without merging**